notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+       "<tr><td>2023-10-25T00:00:00.000Z</td><td>80</td><td>890</td></tr>\n",
+       "<tr><td>2023-10-25T00:00:00.000Z</td><td>30</td><td>872</td></tr>\n",
+       "<tr><td>2023-10-25T00:00:00.000Z</td><td>90</td><td>850</td></tr>\n",
+       "<tr><td>2023-10-25T00:00:00.000Z</td><td>100</td><td>850</td></tr>\n",
+       "<tr><td>2023-10-25T00:00:00.000Z</td><td>10</td><td>624</td></tr>\n",
+       "</table></div>"
+      ],
+      "text/plain": [
+       "<IPython.core.display.HTML object>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "2023-10-26 01:00:06.262543 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 0, 6, 262543)"
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "sql='''\n",
+    "SELECT TIME_FLOOR(c.__time, 'P1D') as _date, \n",
+    "  round( TIMESTAMPDIFF(YEAR, TIME_PARSE(u.dob), CURRENT_TIMESTAMP),-1) as age_group, \n",
+    "  count(distinct c.client_ip) as distint_ips\n",
+    "FROM clicks c \n",
+    "     INNER JOIN users u\n",
+    "        ON c.user_id=u.user_id\n",
+    "WHERE c.__time > CURRENT_TIMESTAMP  - INTERVAL '1' DAY\n",
+    "GROUP BY 1, round( TIMESTAMPDIFF(YEAR, TIME_PARSE(u.dob), CURRENT_TIMESTAMP),-1) \n",
+    "ORDER BY 3 DESC\n",
+    "LIMIT 10\n",
+    "'''\n",
+    "display.sql(sql)\n",
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "93cb2f03-0831-41ae-8df8-865a77280aa2",
+   "metadata": {},
+   "source": [
+    "Run it 20 times..."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "id": "7b97cdfe-f46b-4cd9-8b59-faba4a4c015e",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Results = avg:1182.9867 ms   min:1093.643 ms  max:1455.442 ms\n",
+      "2023-10-26 01:00:29.929201 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 0, 29, 929201)"
+      ]
+     },
+     "execution_count": 14,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "join_results=[]\n",
+    "result = measure_query(sql, 20)\n",
+    "# saving all results for comparison \n",
+    "join_results.append({\"query_type\":\"simple_join\",\"results\":result})\n",
+    "print(result)\n",
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a040384d-3d46-4bce-a084-c020433fa676",
+   "metadata": {},
+   "source": [
+    "### Order of the Joins is Important\n",
+    "\n",
+    "![](assets/fail-scatter-gather.png)\n",
+    "\n",
+    "If you change the order of the joins, it will fail because the subquery will return more than `maxSubQueryRows` which defaults to 100000. "
```

Comment:
Table X returns Y results, table V contains W results.

The configuration parameter `whetherToEverWatchStarTrekVAgain` determines whether peas are tasty.

Run the following cell, where the two parts of the salad have been swapped over, meaning that W and Y will happen.

Reply:
I don't like that as the solution and we shouldn't teach that. Changing that limit is usually because you don't understand the Druid query model and you are likely trying to use for a use case that is not very Druid friendly. I've seen folks in the community change this to 20 million just to later complain about how they are running out of memory on the broker or Peons involved. .... getting off my soapbox now... ;-)

Reply:
On second thought, you are right in that using the queryContext for this purpose would be the right way to use it, but only in exceptions, not where you want it to accommodate anti-pattern SQL. Perhaps a different subject?